pkg/controlplane/apiserver/apis.go

```diff
@@ -83,8 +90,60 @@ func (c *CompletedConfig) GenericStorageProviders(discovery discovery.DiscoveryI
 	}, nil
 }
 
+// StorageReadinessHook implements PostStartHook functionality for checking readiness
+// of underlying storage for registered resources.
+type StorageReadinessHook struct {
+	checks  map[string]func() error
+	timeout time.Duration
+}
+
+// NewStorageReadinessHook created new StorageReadinessHook.
+func NewStorageReadinessHook(timeout time.Duration) *StorageReadinessHook {
+	return &StorageReadinessHook{
+		checks:  make(map[string]func() error),
+		timeout: timeout,
+	}
+}
+
+func (h *StorageReadinessHook) RegisterStorage(gvr metav1.GroupVersionResource, storage rest.StorageWithReadiness) {
```

Comment:
Add comment that this is unsafe? Do we also need to document that the storage.ReadinessCheck will be called at 100ms intervals up to the timeout provided to NewStorageReadinessHook?